### ISSUE TITLE
new XMLSerializer().serializeToString(document); panics

### ISSUE THREAD

**Issue by mateon1** [servo/servo#23130]
```html
<script>new XMLSerializer.serializeToString(document)</script>
```

This causes a ScriptThread panic, `Can't serialize Document node itself`

https://github.com/servo/servo/blob/fcd6beb60819ef2960f70d10e82fa31ba56cc99b/components/script/dom/servoparser/html.rs#L253

Both Chrome and Firefox properly serialize the entire page when `serializeToString` is called on the document.

**Comment by CYBAI** [comment]
As @jdm mentioned in https://github.com/servo/servo/pull/23044#issuecomment-474087096 , I think this will be fixed when servo/html5ever#368 is fixed.

**Comment by jdm** [comment]
Specifically, this would be handled by https://w3c.github.io/DOM-Parsing/#dfn-xml-serialization-algorithm.

**Comment by jdm** [comment]
That being said, fixing the panic here is separate from implementing the full algorithm in the html5ever repository. We should implement the [document serialization algorithm](https://w3c.github.io/DOM-Parsing/#dfn-xml-serializing-a-document-node) by checking for a document in the [SerializationIterator constructor](https://github.com/servo/servo/blob/fcd6beb60819ef2960f70d10e82fa31ba56cc99b/components/script/dom/servoparser/html.rs#L167-L176) and pushing all of the immediate child nodes of the document.

**Comment by jdm** [comment]
If this case is not already tested by tests/wpt/web-platform-tests/domparsing/XMLSerializer-serializeToString.html, then a new testcase should be added.

**Comment by 5h1rU** [comment]
Hello!

would like to contribute here but probably I'll need some assistance since I'm a newbie on the codebase.
@jdm what do you think?

**Comment by jdm** [comment]
@5h1rU I think you should give it a try and ask questions about anything that is unclear :)

**Comment by 5h1rU** [comment]
@highfive assign me

**Comment by 5h1rU** [comment]
After some hours trying to figure out what's going on I have some blockers

- html5ever serializer doesn't have the document serialization algorithm.

Then I change the Document match to:

```rust
NodeTypeId::Document(DocumentTypeId::Document) => {
    let doc = n.downcast::<Node>().unwrap();
    // What else?
},
```
I'm confused about the document serialization algorithm implementation. Not sure where is the best place to put the logic. I tried using the `push_node` from the `SerializationIterator` but no luck.
```rust
iter.push_node(doc);
```
@jdm Do you think is possible to use one of the current existent serializer methods?

In conclusion: I don't know what's the next step in order to return the document serialized.

**Comment by jdm** [comment]
@5h1rU [This comment](https://github.com/servo/servo/issues/23130#issuecomment-478050900) means that we should not modify the match with the panic. However, we should modify the SerializationIterator `new` method to check for a document node, iterate over its children and push each one of them without pushing the actual document node.
